### ISSUE TITLE
exp(p3): MAPPO Phase 2 sweep on default, minimal_specialization, positional_default

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#231]
## Context

This is the experiment the entire recent batch was built to enable. Per project memory:

> Env-side fixes verified working (#197, #198, #199); independent-PPO still plateaus at 15% of optimum on a sanity-check scenario. Next intervention is MAPPO / CTDE.

Since then:

- PR #225 landed MAPPO Option A (centralized critic, default-off, `--centralized-critic` CLI flag wired into `experiments/p3_specialization/train.py::CellConfig`).
- PR #221's empirical work (PR #228) shows **`positional_default` with alpha=0.1 does NOT rescue independent-PPO** — closes only 9.6%/16.1% of the specialist-random gap. That makes two env-side interventions that have empirically failed (the third, #220's per-agent obs differentiation, is awaiting rerun).

The algorithm-side intervention is now the natural next step.

## What to do

On `COMPUTE_HOST_PRIMARY`, paired sweep:

- **Scenarios:** `default`, `minimal_specialization`, `positional_default` (alpha=0.1 — the PR #228 baseline)
- **Algorithm:** IPPO (current behavior) vs MAPPO (`--centralized-critic`)
- **Seeds:** 42, 43, 44 (matching the convention from PRs #221/#225)
- **Iterations:** ≥50, ideally 100 to match #220's protocol
- **Rollout:** 2048 steps per iteration

Total: 3 scenarios × 2 algos × 3 seeds = 18 cells. Parallelize on the Mac Studio.

## Metrics & verdict thresholds

Apply the same ladder as #220:
- ≥50% additional gap closed by MAPPO vs IPPO baseline on at least 2 of 3 scenarios → **MAPPO succeeds**; update the "PPO failure mode" memory note; deprioritize #193 reward shaping
- 25-50% on most scenarios → **MAPPO helps but doesn't solve**; design follow-on (longer training? alpha sweep? combine with reward shaping?)
- <25% across the board → **CTDE alone insufficient**; the bottleneck is deeper; promote #193 to next priority

Per-agent action entropy, pairwise KL (via `experiments/p3_specialization/diagnostics/pairwise_action_kl.py` from PR #227), and the H1/H2 audit reports should all run alongside, especially if a scenario shows MAPPO succeeding — to understand *what changed* about the learned policies.

## References

- PR #225 (MAPPO Option A implementation)
- Issue #220 (per-agent obs validation — pre-registered verdict template)
- PR #228 (positional_default empirical baseline + 3D-obs fixes)
- Issue #199 (minimal_specialization baseline references — random = -96.07/step, specialist = -22.07/step on the canonical reward scale)
- Project memory: "PPO failure mode"

## Acceptance criteria

- [ ] All 18 cells run on the primary host
- [ ] Dated `research_notebook/` entry comparing IPPO vs MAPPO across the 3 scenarios
- [ ] Verdict applied (per the threshold ladder)
- [ ] Project memory updated to reflect the new state
- [ ] PR closing this issue

## Out of scope

- `distance_cost_alpha` sweep on positional_default — defer until baseline MAPPO numbers are in
- Joint MAPPO + reward shaping (#193) — defer based on this experiment's outcome
- Hyperparameter tuning of MAPPO itself — use the same hyperparameters as IPPO for the comparison; tune in a follow-up if MAPPO underperforms

**Comment by rjwalters** [comment]
## Curator enrichment (2026-05-16)

Refreshed against current `main` after PR #225 (MAPPO Option A), PR #228 (positional_default baseline + 3 #216 regression fixes), PR #227 (issue-220 analyzers — including the reusable `pairwise_action_kl.py` diagnostic). The infrastructure to run this sweep is now fully in place.

### Code surface (no changes needed before the sweep)

- `bucket_brigade/training/joint_trainer.py` — `JointPPOTrainer(..., centralized_critic=False)`. When `True`: instantiates `CentralizedCritic` over the global-obs portion (identity tail stripped), shared value baseline broadcast to all N agents, separate Adam optimizer for the critic. Raises `ValueError` if `centralized_critic=True` and `redundancy_coef > 0` (incompatible). Per-agent rewards still drive per-agent advantages — only the value baseline is shared.
- `bucket_brigade/training/networks.py` — `CentralizedCritic` (2-layer MLP trunk, scalar head, mirrors `PolicyNetwork`'s trunk).
- `experiments/p3_specialization/train.py` — `CellConfig(centralized_critic: bool = False)` + `--centralized-critic` argparse flag at line 579. Threaded through to trainer constructor at line 481.
- `experiments/p3_specialization/diagnostics/pairwise_action_kl.py` — landed in PR #227; reads `config.json` + `policies/` from a cell dir and emits `pairwise_action_kl.json`. Reusable as-is.
- Scenario `positional_default` (registered in `bucket_brigade/envs/scenarios_generated.py:402`) bakes `distance_cost_alpha=0.1` into the scenario factory — **no extra CLI flag needed**; passing `--scenario positional_default` is sufficient.

### Hyperparameters (frozen, matching PR #228 protocol for cross-comparability)

```
--num-iterations 100   # full protocol (PR #228 used 50; bump to 100 to match #220)
--rollout-steps 2048
--num-agents 4
--value-coef 0.5
--entropy-coef 0.01
--normalize-returns
--device cpu
```

`--lambda-red 0.0` for every cell (MAPPO is gated against `redundancy_coef > 0`; IPPO arm matches for apples-to-apples).

### 18 exact CLI invocations

```bash
# On COMPUTE_HOST_PRIMARY (robbs-mac-studio), in repo root.
# Single source of truth — drop into a shell script if you prefer:

RUNS_ROOT=experiments/p3_specialization/runs/issue231
SCENARIOS="default minimal_specialization positional_default"
SEEDS="42 43 44"
COMMON="--num-iterations 100 --rollout-steps 2048 --num-agents 4 \
  --value-coef 0.5 --entropy-coef 0.01 --normalize-returns --device cpu \
  --lambda-red 0.0"

# IPPO arm (9 cells)
for S in $SCENARIOS; do for K in $SEEDS; do
  uv run python -m experiments.p3_specialization.train \
    --scenario "$S" --seed "$K" \
    --output-dir "$RUNS_ROOT/ippo/$S/seed_$K" $COMMON
done; done

# MAPPO arm (9 cells) — only difference is the --centralized-critic flag
for S in $SCENARIOS; do for K in $SEEDS; do
  uv run python -m experiments.p3_specialization.train \
    --scenario "$S" --seed "$K" \
    --output-dir "$RUNS_ROOT/mappo/$S/seed_$K" $COMMON \
    --centralized-critic
done; done
```

Each invocation writes `metrics.json`, `config.json`, and `policies/` under its `output-dir`.

### Tmux session layout (6 sessions, parallel across scenarios)

Following PR #227's `issue220-{baseline,treatment}` convention:

```
issue231-ippo-default
issue231-ippo-minspec
issue231-ippo-positional
issue231-mappo-default
issue231-mappo-minspec
issue231-mappo-positional
```

Each session runs **3 cells sequentially** (its 3 seeds for that scenario × algo). 6 sessions × 3 cells = 18 cells. This gives 6-way parallelism across the 8 physical cores on Mac Studio — leaves headroom for the OS and avoids context-switch contention from over-subscription.

Kickoff pattern (one per session — repeat the inner block changing `ARM` and `SCEN_SHORT`):

```bash
ssh "$COMPUTE_HOST_PRIMARY"
export PATH="/opt/homebrew/bin:$HOME/.cargo/bin:$PATH"  # non-interactive SSH PATH quirk
cd ~/GitHub/bucket-brigade  # or ~/bucket-brigade — check both
git pull
uv sync --extra rl

ARM=ippo            # or "mappo"
SCEN=default        # or "minimal_specialization" or "positional_default"
SCEN_SHORT=default  # or "minspec" or "positional"
FLAG=""             # or "--centralized-critic" for the mappo arm

tmux new -d -s "issue231-$ARM-$SCEN_SHORT" "
  export PATH='/opt/homebrew/bin:\$HOME/.cargo/bin:\$PATH';
  cd ~/GitHub/bucket-brigade &&
  for K in 42 43 44; do
    uv run python -m experiments.p3_specialization.train \
      --scenario $SCEN --seed \$K \
      --output-dir experiments/p3_specialization/runs/issue231/$ARM/$SCEN/seed_\$K \
      --num-iterations 100 --rollout-steps 2048 --num-agents 4 \
      --value-coef 0.5 --entropy-coef 0.01 --normalize-returns --device cpu \
      --lambda-red 0.0 $FLAG \
      2>&1 | tee -a experiments/p3_specialization/runs/issue231/$ARM-$SCEN_SHORT.log;
  done
"
```

### Parallelism + wall-clock

- **Per-cell cost**: PR #227 reports ~3 min/cell at 50 iters/2048 steps on Mac Studio. At 100 iters that's ~6 min/cell.
- **Per session** (3 cells sequentially): ~18 min wall-clock.
- **6 sessions in parallel** on 8 physical cores: CPU-bound, no cross-session memory pressure (each cell is a single small MLP + 4-agent rollout). Expect ~20–25 min wall-clock for the full 18-cell sweep, allowing for some thermal throttling.
- **Memory**: each Python process is well under 1 GB. 6 × ~1 GB ≪ Mac Studio's typical 32–64 GB.
- **Verify before kickoff**: `ssh -o BatchMode=yes -o ConnectTimeout=5 $COMPUTE_HOST_PRIMARY echo ok`.

### Results directory layout

```
experiments/p3_specialization/runs/issue231/
├── ippo/
│   ├── default/seed_{42,43,44}/{config,metrics}.json + policies/
│   ├── minimal_specialization/seed_{42,43,44}/...
│   └── positional_default/seed_{42,43,44}/...
├── mappo/
│   ├── default/seed_{42,43,44}/...
│   ├── minimal_specialization/seed_{42,43,44}/...
│   └── positional_default/seed_{42,43,44}/...
└── *.log  (per-session tee'd output)
```

Per-cell pairwise KL (run after training; cheap):

```bash
for ARM in ippo mappo; do for S in default minimal_specialization positional_default; do
  for K in 42 43 44; do
    uv run python experiments/p3_specialization/diagnostics/pairwise_action_kl.py \
      --cell "experiments/p3_specialization/runs/issue231/$ARM/$S/seed_$K"
  done
done; done
```

### Analyzer: `experiments/p3_specialization/analyze_231.py` (to create)

Model after `analyze_220.py` (which already implements the right shape — read it, then adapt). Key differences from `analyze_220.py`:

1. **Three scenarios**, not two: `default`, `minimal_specialization`, `positional_default`.
2. **Two arms named `ippo` / `mappo`** (not `baseline` / `treatment`). The cell-dir path is `runs/issue231/{arm}/{scenario}/seed_{N}/` — note no `lambda_0e0` subdir (issue231 fixes lambda at 0; we drop the lambda subdir from the layout above to keep paths shorter).
3. **Per-scenario gap-closure computation**. `analyze_220.py` only computes gap-closed on `minimal_specialization`. Issue #231 wants it on **all three scenarios**. Need scenario-specific (random, specialist) reference pairs:
   - `minimal_specialization`: random = −96.07, specialist = −22.07 (from PR #228 `baselines.json`; also referenced in #199).
   - `positional_default`: pull from `experiments/p3_specialization/diagnostics/results/issue221_positional/baselines.json` (random and specialist per-step values).
   - `default`: PR #228's notebook entry has these — load from the same `baselines.json` or hard-code with citation. (Per #199 measurements: random ≈ −96.07/step, specialist ≈ −22.07/step on the canonical reward scale — verify against current refs before running.)
4. **Per-scenario verdict** plus **headline verdict**:
   - For each scenario, compute `additional_gap_closed = gap_mappo - gap_ippo` (treating IPPO as the baseline).
   - For each scenario, verdict tier: `≥0.50` → MAPPO succeeds; `0.25–0.50` → partial; `<0.25` → insufficient.
   - Headline verdict = `MAPPO_SUCCEEDS` iff ≥2 of 3 scenarios at tier-1 (matching issue body); `PARTIAL` iff ≥2 scenarios at tier-1-or-2 but headline not met; `INSUFFICIENT` otherwise.
5. **Edge-case verdict refinements** (extend the ladder from the issue body):
   - **Split outcome** (e.g., MAPPO passes `default` but fails `minimal_specialization`): emit `PARTIAL_SCENARIO_DEPENDENT` and list the per-scenario verdicts in the summary. This is informative — `default` passing while `minimal_specialization` fails would suggest the centralized critic helps coordination but doesn't break role-collapse, pointing follow-up at reward shaping (#193) for hard scenarios specifically.
   - **MAPPO regresses** (`additional_gap_closed < -0.10` on any scenario): emit `MAPPO_HARMFUL_ON_$SCENARIO` so we don't silently bury negative transfer.
   - **All three at tier-2** (25–50%): emit `MAPPO_HELPS_GLOBALLY_NOT_DECISIVE` — different from a 2-of-3 split, points to "longer training or hyperparameter tuning" rather than "abandon CTDE".
6. **Output**: `experiments/p3_specialization/diagnostics/results/issue231_mappo/summary.{md,json}` with:
   - Team-reward table (3 scenarios × 2 arms × per-seed + mean ± std).
   - Per-agent action entropy + entropy spread + pairwise KL (3 × 2 grid).
   - Gap-closed table (3 × 2 grid + delta column + per-scenario tier).
   - Headline verdict block.

### Pre-registered verdict (consolidated, with edge cases)

| Outcome | Trigger | Action |
|---|---|---|
| **MAPPO succeeds** | ≥2 of 3 scenarios with `mappo_gap - ippo_gap ≥ 0.50` | Update "PPO failure mode" memory; deprioritize #193 reward shaping; promote follow-ons (alpha sweep, longer training) |
| **MAPPO helps globally** | All 3 scenarios at `mappo_gap - ippo_gap ∈ [0.25, 0.50)` | Design follow-on: longer training, alpha sweep, or combine MAPPO + reward shaping |
| **Scenario-dependent** | ≥1 scenario passes, ≥1 fails | Emit per-scenario verdicts; the structural answer depends on which scenario family wins — points to scenario-specific follow-ups |
| **MAPPO insufficient** | All 3 scenarios at `mappo_gap - ippo_gap < 0.25` | CTDE alone insufficient; promote #193 to next priority; consider QMIX or other CTDE-with-mixing |
| **MAPPO harmful** | Any scenario with `mappo_gap - ippo_gap < −0.10` | Block headline verdict; investigate (critic instability? value-target mismatch?) before declaring on #208 line of work |

### Pre-flight checks (run once before kickoff)

```bash
# 1. Verify host reachable
ssh -o BatchMode=yes -o ConnectTimeout=5 $COMPUTE_HOST_PRIMARY echo ok

# 2. Smoke a single cell on remote (5 iters, ~30s) to confirm --centralized-critic still works on main
ssh "$COMPUTE_HOST_PRIMARY" "
  export PATH='/opt/homebrew/bin:\$HOME/.cargo/bin:\$PATH';
  cd ~/GitHub/bucket-brigade && git pull && uv sync --extra rl &&
  uv run python -m experiments.p3_specialization.train \
    --scenario minimal_specialization --seed 42 \
    --output-dir /tmp/issue231_smoke --num-iterations 5 --rollout-steps 512 \
    --num-agents 4 --value-coef 0.5 --entropy-coef 0.01 --normalize-returns \
    --device cpu --lambda-red 0.0 --centralized-critic
"

# 3. Confirm `positional_default` scenario factory exists in current HEAD on remote
ssh "$COMPUTE_HOST_PRIMARY" "cd ~/GitHub/bucket-brigade && \
  grep -q positional_default bucket_brigade/envs/scenarios_generated.py && echo ok"
```

### Updated acceptance criteria

- [ ] All 18 cells run on `$COMPUTE_HOST_PRIMARY` and produce `metrics.json` + `config.json` + `policies/`
- [ ] `pairwise_action_kl.json` produced for all 18 cells via `pairwise_action_kl.py`
- [ ] `experiments/p3_specialization/analyze_231.py` created and produces `summary.{md,json}` under `experiments/p3_specialization/diagnostics/results/issue231_mappo/`
- [ ] Dated `research_notebook/` entry comparing IPPO vs MAPPO across the 3 scenarios, including per-scenario gap-closure table + headline verdict + edge-case verdict if applicable
- [ ] Verdict applied per the threshold ladder above (including edge-case refinements)
- [ ] Project memory ("PPO failure mode" note) updated to reflect the new state
- [ ] PR closes this issue

### References (additions)

- PR #227 — established the `pairwise_action_kl.py` diagnostic and `analyze_220.py` template that `analyze_231.py` should mirror.
- PR #228 — established the 50-iter/100-iter PPO hyperparameter convention (`--value-coef 0.5 --entropy-coef 0.01 --normalize-returns`); also fixed three latent regressions from PR #216 that would have blocked this sweep.
